src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2014, Kaazing Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kaazing.netx.ws.internal.util;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+// ### TODO: 1. happens-before relationship should be established between the thread that releases the lock and the next thread
+//              that acquires the lock.
+//           2. Evaluate using Martin's sun.misc.Unsafe based AtomicSequence implementation.
+public class OptimisticReentrantLock {
```

Comment:
Understood.

You may want to either consider:
  o a back-off strategy in the spin loops to make sure other threads are given more opportunity to get scheduled, make progress, and let the spin loop exit (lock-free),
OR
  o using `tryLock` instead of `lock` in the implementations, and letting the `read` or `write` calls return immediately without spinning (wait-free)